### PR TITLE
Pull in fairscale.nn.Pipe into PyTorch.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44090 Pull in fairscale.nn.Pipe into PyTorch.**

This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

### CODE REVIEW

**Review by msbaines — COMMENTED** [review]
lgtm

File: LICENSE [hunk 1]
```diff
@@ -16,23 +16,26 @@ Copyright (c) 2016-present, Facebook Inc. All rights reserved.
...
 Copyright (c) 2015 Google Inc.
 All rights reserved.
- 
+
 All contributions by Yangqing Jia:
 Copyright (c) 2015 Yangqing Jia
 All rights reserved.
- 
+
+All contributions by Kakao Brain:
```
Comment (pritamdamania87) [inline LICENSE:28]: license

File: NOTICE [hunk 2]
```diff
@@ -22,6 +22,9 @@ All contributions by Yangqing Jia:
 Copyright (c) 2015 Yangqing Jia
 All rights reserved.
 
+All contributions by Kakao Brain:
```
Comment (pritamdamania87) [inline NOTICE:25]: notice

File: test/distributed/_pipeline/sync/conftest.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
```
Comment (pritamdamania87) [inline test/distributed/_pipeline/sync/conftest.py:1]: license header

**Comment by rohan-varma** [comment]
CI issue for Mac/Windows looks legit:
```
Oct 07 05:00:59 Traceback (most recent call last):
Oct 07 05:00:59   File "test/run_test.py", line 830, in <module>
Oct 07 05:00:59     main()
Oct 07 05:00:59   File "test/run_test.py", line 813, in main
Oct 07 05:00:59     raise RuntimeError(err_message)
Oct 07 05:00:59 RuntimeError: distributed/_pipeline/sync/skip/test_api failed!
```

File: NOTICE [hunk 4]
```diff
@@ -22,6 +22,9 @@ All contributions by Yangqing Jia:
...
+Copyright 2019-2020 Kakao Brain
```
Comment (mrshenli) [inline NOTICE:26]: I assume the legal team has reviewed these license updates and they are OK with it?

File: test/run_test.py [hunk 5]
```diff
@@ -95,6 +96,54 @@
...
+    'distributed/_pipeline/sync/test_deferred_batch_norm',
+    'distributed/_pipeline/sync/test_dependency',
+    'distributed/_pipeline/sync/test_inplace',
+    'distributed/_pipeline/sync/test_microbatch',
+    'distributed/_pipeline/sync/test_phony',
+    'distributed/_pipeline/sync/test_pipe',
+    'distributed/_pipeline/sync/test_pipeline',
+    'distributed/_pipeline/sync/test_stream',
+    'distributed/_pipeline/sync/test_transparency',
+    'distributed/_pipeline/sync/test_worker',
```
Comment (mrshenli) [inline test/run_test.py:120]: how much time does this add to general CI tests? If it is a lot, shall we consider adding it to slow tests for now to minimize the impact on other PRs?

**Review by mrshenli — APPROVED** [review]
Folder structure looks OK to me. 

Since we imported the code as-is from torch-gpipe, the current copyright makes sense. In the future, if we modified say 100% of one file, do we need to update the copyright in that file and the corresponding folder accordingly?

File: NOTICE (hunk 4 above)
Comment (pritamdamania87) [reply NOTICE:26]: Yes, this was added as per their guidance and I believe they mentioned they have a job that verifies licensing changes once things land. So they might reach out later if needed.

File: test/run_test.py (hunk 5 above)
Comment (pritamdamania87) [reply test/run_test.py:120]: Most of the tests didn't take too long, but I still added all of them to SLOW_TESTS to be safe.

**Pull request closed without merging**